* Preprocess
** Sort
grep miR-17- Predicted_Target_Locations.default_predictions.hg19.bed |cut -f1-3,6 |sort -k1,1 -k2,2n > mir17.bed
grep miR-221- Predicted_Target_Locations.default_predictions.hg19.bed |cut -f1-3,6 |sort -k1,1 -k2,2n > mir221.bed
** liftover ucsc
sort -k1,1 -k2,2n mir221_liftover.bed |perl -wlane 'print join("\t",join("\t",@F[0..2]),"name",255,$F[3])' > mir221.bed
sort -k1,1 -k2,2n mir17_liftover.bed |perl -wlane 'print join("\t",join("\t",@F[0..2]),"name",255,$F[3])' > mir17.bed
** Find closest pairs
closestBed -D a -d -io -s -a mir17.bed -b mir221.bed |grep -v -P "\-1\t\-1" > Closest_17_221.bed
python -c "import sys;[sys.stdout.write('\t'.join(line.split('\t')[:])) for line in sys.stdin if (int(line.split('\t')[-1]) <= 200 and int(line.split('\t')[-1]) >= -200)]" < Closest_17_221.bed > Closest_17_221_200.bed
** Annotate
rsync -auv jack:/scratch2/fall/Data/Annotation/Homo_sapiens.GRCh38.95.gtf.gz .
perl ~/Scripts/Workflowscripts/Universal/AnnotateBed.pl -b Closest_17_221_200.bed -a Homo_sapiens.GRCh38.95.gff3.gz -w ON --specific gene > Annotated_17_221_closest.bed
grep ENSG Annotated_17_221_closest.bed | perl -lane 'print join("\t",$F[0],$F[15],$F[16],$F[14],".",$F[5])' |sort -V |uniq > TargetGenes_17_221.bed
** Create fastas
fastaFromBed -fi hg38.extended.fa -bed TargetGenes_17_221.bed -name+ -s |gzip > 17_221_targets.fa.gz
** Convert to local coords
perl -lane '{if ($F[5] eq "\+" or $F[5] eq "u" or $F[5] eq "." or $F[5] > 0){$start = $F[1]-$F[15];}else{$start = $F[16]-$F[2]} if ($start >= 0 ){ $end = $start+9; print join("\t",$F[0],$start,$end,@F[3..5])}}' Annotated_17_221_closest.bed |sort -k1,1 -k2,2n -u > 17_local_constraints.bed
perl -lane '{if ($F[5] eq "\+" or $F[5] eq "u" or $F[5] eq "." or $F[5] > 0){$start = $F[8]-$F[15];}else{$start = $F[16]-$F[9]} if ($start >= 0 ){ $end = $start+9; print join("\t",$F[0],$start,$end,@F[3..5])}}' Annotated_17_221_closest.bed |sort -k1,1 -k2,2n -u > 221_local_constraints.bed
** Create paired constraint
perl -lane 'if ($F[0]=~ /chr/ && $F[7] =~ /chr/){print join("\t",$F[0],$F[1]-$F[15],$F[2]-$F[15],@F[3..7],$F[8]-$F[15],$F[9]-$F[15],@F[10..$#F])}' Annotated_17_221_closest.bed |perl -lane 'if($F[1]>=0 && $F[2] >= 0 && $F[8] >= 0 && $F[9] >= 0){print $_}' |grep ENSG |sort -k1,1 -k2,2n -u > Local_Paired_17_221_constraints.bed
** Run on miRNA 17/221
*** PLFold
**** Calculate
 ~/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintPLFold.py -s 17_221_targets.fa.gz -w 250 -r raw -n unpaired -p paired -x 17_local_constraints.bed -o Plcons_17 -z 4
 ~/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintPLFold.py -s 17_221_targets.fa.gz -w 250 -r raw -n unpaired -p paired -x 221_local_constraints.bed -o Plcons_221 -z 4 --loglevel WARNING
**** Collect
cd Plcons_17
python ~/Projects/INPROGRESS/RIssmed/RIssmed/CollectConsResults.py -u 1 -z 1 -g ../TargetGenes_17_221.bed -b 0,1 -c 0 -p 250,60 -o Collection
cd Plcons_221
python ~/Projects/INPROGRESS/RIssmed/RIssmed/CollectConsResults.py -u 1 -z 1 -g ../TargetGenes_17_221.bed -b 0,1 -c 0 -p 250,60 -o Collection
*** Fold
**** single (bringt hier nix)
python /home/fall/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintFold.py -s 17_221_targets.fa.gz -w 100 -n unpaired -p paired -r add -x 17_local_constraints.bed -i TargetGenes_17_221.bed -z 4 -o Window_17
python /home/fall/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintFold.py -s 17_221_targets.fa.gz -w 250 -n unpaired -p paired -r add -x 221_local_constraints.bed -i TargetGenes_17_221.bed -z 4 -o Window_221
**** paired
python /home/fall/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintFold.py -s 17_221_targets.fa.gz -w 100 -l 90 -n unpaired -p paired -r add -x Local_Paired_17_221_constraints.bed -i TargetGenes_17_221.bed -z 4 -o Window_paired
**** Collect
cd Window_paired
python ~/Projects/INPROGRESS/RIssmed/RIssmed/CollectWindowResults.py -z 1 -g ../TargetGenes_17_221.bed -b " -1,1" -p 100,90 -o Collection
* Run on FUBP and U2AF65
cd /scr/k70san2/fall/Constraints/Mainz/ReStart_U2AF65_FUBP
ca rissmed
#rsync -auv ../U2AF65_FUBP_NEW65/*.fa*gz .
** Reprocess, for Preprocess of old files see ~/Projects/INPROGRESS/FoldConstraints/Cooperativity_Mainz/workflow.org
*** Sort
sort -k1,1 -k2,2n U2AF65_peaks_reproducible_3CL.bed |uniq > Sorted_65_peaks_hg38.bed
sort -k1,1 -k2,2n FUBP1_peaks_reproducible_3CL.bed |uniq > Sorted_Fu_peaks_hg38.bed
*** Find closest pairs
closestBed -D a -d -io -s -a Sorted_65_peaks_hg38.bed -b Sorted_Fu_peaks_hg38.bed |grep -v -P "\-1\t\-1" > Closest_65_Fu.bed
python -c "import sys;[sys.stdout.write('\t'.join(line.split('\t')[:])) for line in sys.stdin if (int(line.split('\t')[12]) <= 150 and int(line.split('\t')[12]) >= -150)]" < Closest_65_Fu.bed > Closest_65_Fu_150.bed
*** Merge and annotate
rsync -auv jack:/scratch2/fall/Data/Annotation/Homo_sapiens.GRCh38.88.gtf.gz .
perl ~/Work/Projects/Workflows/scripts/AnnotateBed.pl -b Closest_65_Fu_150.bed -a Homo_sapiens.GRCh38.88.gtf.gz -w ON --specific gene |sort -V|uniq|perl -wlane 'print "chr".$_' |gzip > Annotated_65_Fu_closest.bed.gz
grep -w gene <(zcat Annotated_65_Fu_closest.bed.gz) |sort -V |gzip > Annotated_65_Fu_closest_gene_overlap.bed.gz
perl -wlane 'print join("\t",$F[0],$F[15],$F[16],$F[14],".",$F[5])' <(zcat Annotated_65_Fu_closest_gene_overlap.bed.gz)|sort -V |uniq> TargetGenes_65_Fu.bed
perl -wlane 'print join("\t",@F[0..5])' <(zcat Annotated_65_Fu_closest.bed.gz)|sort -V |uniq > 65_peaks_for_folding.bed
perl -wlane 'print join("\t",@F[7..9],$F[3],@F[11..12])' <(zcat Annotated_65_Fu_closest.bed.gz)|sort -V |uniq > Fu_peaks_for_folding.bed
*** Get fasta with strand info!!!
fastaFromBed -fi hg38.extended.fa -bed TargetGenes_65_Fu.bed -name+ -s |gzip > 65_Fu_targets.fa.gz
*** TODO Get local coords of Fu binding on 65 extension, closestBed -D a gives us dist in regard to a, as we consider strand in fasta this is important
perl -lane '{if ($F[5] eq "\+" or $F[5] eq "u" or $F[5] eq "." or $F[5] > 0){$start = $F[1]-$F[15]+1;}else{$start = $F[16]-$F[2]+1} if ($start > 0 ){ $end = $start+9; print join("\t",$F[0],$start,$end,@F[3..5])}}' Annotated_65_Fu_closest_gene_overlap.bed |sort -k1,1 -k2,2n > 65_local_constraints.bed
perl -lane '{if ($F[5] eq "\+" or $F[5] eq "u" or $F[5] eq "." or $F[5] > 0){$start = $F[8]-$F[15]+1;}else{$start = $F[16]-$F[9]+1} if ($start > 0 ){ $end = $start+9; print join("\t",$F[0],$start,$end,@F[3..5])}}' Annotated_65_Fu_closest_gene_overlap.bed |sort -k1,1 -k2,2n > Fu_local_constraints.bed
** Run prob folding
*** on k61
 ca rissmed
 ~/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintPLFold.py -s 65_Fu_targets.fa.gz -w 250 -l 150 -u 5 -g TargetGenes_65_Fu.bed -r raw -n unpaired -p paired -x 65_peaks_for_folding.bed -o 65_local_reproduce -z 30 --loglevel WARNING -m 2
 ~/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintPLFold.py -s 65_Fu_targets.fa.gz -w 250 -l 150 -u 5 -g TargetGenes_65_Fu.bed -r raw -n unpaired -p paired -x Fu_peaks_for_folding.bed -o Fu_local_reproduce -z 30 --loglevel WARNING -m 2
**** Collect
 Need to get rid of direct overlaps and then calculate zscore and stuff!
***** 65
 cd 65_screen
 rsync -auv /scr/jack2/fall/Projects/Cooperativity/U2AF65_FUBP/TargetGenes_65_Fu.bed .
 python ~/Projects/INPROGRESS/RIssmed/RIssmed/CollectConsResults.py -u 5 -z 8 -g ../TargetGenes_65_Fu.bed -b-1,1 -c 0 -p 250,160 -o Collection_65
 cd Collection_65
 zcat Collection_unpaired.bed.gz|perl -wlane 'if ($F[-1]<-2 || $F[-1] > 2){print}'|gzip > Sig_Collection_unpaired.bed.gz
 zcat Collection_paired.bed.gz|perl -wlane 'if ($F[-1]<-2 || $F[-1] > 2){print}'|gzip > Sig_Collection_paired.bed.gz
***** Fu
 cd Fu_screen
 rsync -auv /scr/jack2/fall/Projects/Cooperativity/U2AF65_FUBP/TargetGenes_65_Fu.bed .
 python ~/Projects/INPROGRESS/RIssmed/RIssmed/CollectConsResults.py -u 5 -z 8 -g ../TargetGenes_65_Fu.bed -b-1,1 -c 0 -p 250,160 -o Collection_Fu
 zcat Collection_unpaired.bed.gz|perl -wlane 'if ($F[-1]<-2 || $F[-1] > 2){print}'|gzip > Sig_Collection_unpaired.bed.gz
 zcat Collection_paired.bed.gz|perl -wlane 'if ($F[-1]<-2 || $F[-1] > 2){print}'|gzip > Sig_Collection_paired.bed.gz
***** Intersect
 intersectBed -s -a Collection_65/Sig_Collection_unpaired.bed.gz -b Fu_peaks_for_folding.bed -wa -wb |gzip > Intersect_sig65_Fu.bed.gz
 intersectBed -s -a Collection_65/Sig_Collection_unpaired.bed.gz -b 65_peaks_for_folding.bed -wa -wb |gzip > Intersect_sig65_65.bed.gz
 intersectBed -s -a Collection_Fu/Sig_Collection_unpaired.bed.gz -b ../../Fu_peaks_for_folding.bed -wa -wb |gzip > Intersect_sigFu_Fu.bed.gz
 intersectBed -s -a Collection_Fu/Sig_Collection_unpaired.bed.gz -b ../../65_peaks_for_folding.bed -wa -wb |gzip > Intersect_sigFu_65.bed.gz
*** on slurm
ssh k70
cd /scr/k70san2/fall/Constraints/Mainz/Reproduced_U2AF65_FUBP
ca rissmed
sbatch rissmed_65.sh
sbatch rissmed_fu.sh
** Window folding
*** Preprocess
cat Annotated_65_Fu_closest_gene_overlap.bed |perl -lane 'if($F[1]>=0 && $F[2] >= 0 && $F[8] >= 0 && $F[9] >= 0){print $_}' |grep ENSG |sort -k1,1 -k2,2n -u > Paired_65_Fu_constraints.bed
cut -f 1-6,8-13 Paired_65_Fu_constraints.bed > Minimal_paired_65_Fu_constraints.bed
**** Old
perl -lane 'if ($F[0]=~ /chr/ && $F[7] =~ /chr/){print join("\t",$F[0],$F[1]-$F[15],$F[2]-$F[15],@F[3..7],$F[8]-$F[15],$F[9]-$F[15],@F[10..$#F])}' Annotated_65_Fu_closest_gene_overlap.bed |perl -lane 'if($F[1]>=0 && $F[2] >= 0 && $F[8] >= 0 && $F[9] >= 0){print $_}' |grep ENSG |sort -k1,1 -k2,2n -u > Local_Paired_65_Fu_constraints.bed
cut -f 1-6,8-13 Local_Paired_65_Fu_constraints.bed > Minimal_local_paired_65_Fu_constraints.bed
*** Run
python ~/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintFold.py -s 65_Fu_targets.fa.gz -w 120 -l 80 -n unpaired -p paired -r add -x Minimal_paired_65_Fu_constraints.bed -i TargetGenes_65_Fu.bed -z 25 -o Window_paired_65_FU
**** Old
python ~/Projects/INPROGRESS/RIssmed/RIssmed/ConstraintFold.py -s 65_Fu_targets.fa.gz -w 120 -l 80 -n unpaired -p paired -r add -x Minimal_local_paired_65_Fu_constraints.bed -i TargetGenes_65_Fu.bed -z 25 -o Window_paired_65_FU
*** Collect
cd Window_paired_65_Fu
python ~/Projects/INPROGRESS/RIssmed/RIssmed/CollectWindowResults.py -z 4 -g ../TargetGenes_65_Fu.bed -b-1,1 -p 120,80 -o Collection_Window_paired
